File: Baekjoon/Breadth_First_Search/1325.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

int N, M;
vector<vector<int> > graph(10001);
vector<int> res(10001);

int bfs(int start) {
	queue<int> q;
	q.push(start);

	vector<bool> visited(N+1, false);
	visited[start] = true;

	int count = 0;
	while (!q.empty()) {
		int v = q.front();
		q.pop();

		for (int i : graph[v]) { // for i in graph[v]
			if (!visited[i]) {
				q.push(i);
				visited[i] = true;
				count += 1;
			}
		}
	}
	return count;
}

int main() {
	cin >> N >> M;

	graph.resize(N+1);
	for (int i = 0; i < M; i++) {
		int a, b;
		cin >> a >> b; // a가 b를 신뢰
		graph[b].push_back(a);
	}

	res.resize(N+1);
	for (int i = 1; i <= N; i++) {
		res[i] = bfs(i);
	}

	// 두 수 비교 -> max(a, b)
	// 벡터에서의 최댓값 -> *max_element(범위)
	int max_val = *max_element(res.begin(), res.end());
	for (int i = 1; i <= N; i++) {
		if (res[i] == max_val) {
			cout << i << " ";
		}
	}

	return 0;
}

File: Baekjoon/Breadth_First_Search/1325.py
from collections import deque
import sys

input = sys.stdin.readline

N, M = map(int, input().split()) # 노드, 간선

graph = [[] for _ in range(N+1)]
for _ in range(M):
	a, b = map(int, input().split()) # a가 b를 신뢰
	graph[b].append(a)

def bfs(start):
	queue = deque()
	queue.append(start)

	visited = [False] * (N+1)
	visited[start] = True

	count = 0
	while queue:
		v = queue.popleft()

		for i in graph[v]:
			if visited[i] == False:
				queue.append(i)
				visited[i] = True
				count += 1
	return count

res = []
for i in range(1, N+1):
	res.append(bfs(i))

for i in range(N):
	if max(res) == res[i]:
		print(i+1, end=" ")
